cmd/metrics-v2.go

```diff
@@ -0,0 +1,1186 @@
+	ch <- c.desc
+}
+
+// populateAndPublish populates and then publishes the metrics generated by the generator function.
+func populateAndPublish(generatorFn func() []MetricsGenerator, publish func(m Metric) bool) {
+	generators := generatorFn()
+	for _, g := range generators {
+		metrics := g()
+		metrics.initialize(GlobalContext, &metrics)
+		for _, metric := range metrics.Metrics {
+			if !publish(metric) {
+				return
+			}
+		}
+	}
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+func (c *minioCollectorV2) Collect(ch chan<- prometheus.Metric) {
+
+	// Expose MinIO's version information
+	minioVersionInfo.WithLabelValues(Version, CommitID).Set(1.0)
+
+	populateAndPublish(c.generator, func(metric Metric) bool {
+		labels, values := getOrderedLabelValueArrays(metric.VariableLabels)
+		values = append(values, GetLocalPeer(globalEndpoints))
+		labels = append(labels, serverName)
+
+		if metric.Description.Type == histogramMetric {
+			if metric.Histogram == nil {
+				return true
+			}
+			for k, v := range metric.Histogram {
+				labels = append(labels, metric.HistogramBucketLabel)
+				values = append(values, k)
+				ch <- prometheus.MustNewConstMetric(
+					prometheus.NewDesc(
+						prometheus.BuildFQName(string(metric.Description.Namespace),
+							string(metric.Description.Subsystem),
+							string(metric.Description.Name)),
+						metric.Description.Help,
+						labels,
+						metric.StaticLabels,
+					),
+					prometheus.GaugeValue,
+					float64(v),
+					values...)
+			}
+			return true
+		}
+
+		metricType := prometheus.GaugeValue
+		switch metric.Description.Type {
+		case counterMetric:
+			metricType = prometheus.CounterValue
+		}
+		ch <- prometheus.MustNewConstMetric(
+			prometheus.NewDesc(
+				prometheus.BuildFQName(string(metric.Description.Namespace),
+					string(metric.Description.Subsystem),
+					string(metric.Description.Name)),
+				metric.Description.Help,
+				labels,
+				metric.StaticLabels,
+			),
+			metricType,
+			metric.Value,
+			values...)
+		return true
+	})
+}
+
+func getOrderedLabelValueArrays(labelsWithValue map[string]string) (labels, values []string) {
+	labels = make([]string, 0)
+	values = make([]string, 0)
+	for l, v := range labelsWithValue {
+		labels = append(labels, l)
+		values = append(values, v)
+	}
+	return
+}
+
+// newMinioCollectorV2 describes the collector
+// and returns reference of minioCollector for version 2
+// It creates the Prometheus Description which is used
+// to define Metric and  help string
+func newMinioCollectorV2(generator func() []MetricsGenerator) *minioCollectorV2 {
+	return &minioCollectorV2{
+		generator: generator,
+		desc:      prometheus.NewDesc("minio_stats", "Statistics exposed by MinIO server", nil, nil),
+	}
+}
+
+func metricsServerHandler() http.Handler {
+
+	registry := prometheus.NewRegistry()
+
+	// Report all other metrics
+	err := registry.Register(newMinioClusterCollector())
+	logger.LogIf(GlobalContext, err)
```

Comment:
we could perhaps panic here like? `if err != nil { logger.CriticalIf(GlobalContext, err) } `